Added support for Raspberry Pi Zero W

NoobieDog opened the pull request:
Ive successfully managed to get the nexmon mod working on RasPi Zero W. 

Any issues please let me know

JBardey commented:
Hi there again, 

The double pipe 'or' operator doesn't work in the version of make that ships on the pi zero. It ignores the second condition which causes nexutil/libnexio to look for the android sdk, and causes the patching to fail with 'this isn't a raspberry pi' messages.

I replaced:

ifeq ($(shell uname -m),armv7l || armv6l)

with

ifeq ($(shell uname -m), $(filter $(shell uname -m), armv7l armv6l))

in patches/bcm43438/7_45_41_26/nexmon/Makefile and did the same to utilities/libnexio/Makefile and utilities/nexutil/Makefile

it all seems to build and I got monitor mode working (at least I saw a lot of stations testing airodump).

Sorry for leaving comments, I don't know how to work git or I'd submit a pull request of my own!

NoobieDog commented:
No thanks for the comments, in all honesty i threw the files together offline, just remembering what i had changed, but i changed armv7l to armv6l..

Thanks for the input and helping. This is what the community is about! :D

matthiasseemoo commented:
Thank you very much for updating the build support for the rpi zero w. I just added a dependency to build libnexio before building nexutil. That was a long awaited little change in the Makefile, but I always forgot to insert this change.
As I currently have no rpi3 with me, could you please check, if everything compiles with the new makefile and then adjust your changes to the nexutil related instruction in the readme.

JBardey commented:
I was having some issues with this after rebooting my Pi Zero W (nexutil seg faults, no wireless device), so starting fresh and will report back.

I will be able to test with a Pi 3 next week.

NoobieDog commented:
@JBardey i ran mine a few times and it worked fine, the only issue i saw was using slow sd cards

NoobieDog commented:
@JBardey Did you get it sorted, i checked mine today after a "break" and worked fine.

JBardey commented:
Yes! I think I got my drivers muddled up (d'oh!). 

Works fine, been using it a couple days. Just needs the steps from the Pi3 to unload/reload the driver after boot. 

Sent from my iPhone

> On 13 Mar 2017, at 17:17, NoobieDog <notifications@github.com> wrote:
> 
> @JBardey Did you get it sorted, i checked mine today after a "break" and worked fine.
> 
> ―
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

fangfufu commented:
Does the armv7l actually work for armv6l?

NoobieDog commented:
@fangfufu, so far it has worked well. i haven't had any issues with the 7l Toolchain on the 6L. 

@JBardey Yeah i had to do that too, ill throw it in to pull request to to the reload on boot :D

fangfufu commented:
@NoobieDog, you didn't include a patch to ``utilities/nexutil/Makefile``.

NoobieDog commented:
@fangfufu for what? 

https://github.com/seemoo-lab/nexmon/pull/55/files#diff-a815e52ca4ebb22d53c8d8785fa0c2af

fangfufu commented:
> in patches/bcm43438/7_45_41_26/nexmon/Makefile and did the same to utilities/libnexio/Makefile and utilities/nexutil/Makefile

You linked me the patch for ``utilities/libnexio/Makefile``, I would like the patch for ``utilities/nexutil/Makefile`` pretty please. :)

``nexutil`` is the only thing that's not compiling on my machine. 

    ~/nexmon/utilities/nexutil# make
    cd ../libargp && make
    make[1]: Entering directory '/root/nexmon/utilities/libargp'
    /ndk-build NDK_APPLICATION_MK=`pwd`/Application.mk NDK_APP_OUT=. TARGET_PLATFORM=android-21
    /bin/sh: 1: /ndk-build: not found
    Makefile:2: recipe for target 'libs/armeabi/libargp.a' failed
    make[1]: *** [libs/armeabi/libargp.a] Error 127
    make[1]: Leaving directory '/root/nexmon/utilities/libargp'
    Makefile:19: recipe for target '../libargp/local/armeabi/libargp.a' failed
    make: *** [../libargp/local/armeabi/libargp.a] Error 2

NoobieDog commented:
ah yeah, no probs, ill sort that tonight and release a test version after work  💃

fangfufu commented:
I got it working after removing the if statements in the Makefiles. :+1:

NoobieDog commented:
@JBardey So today i started getting issues haha, Unloading and Reloading brcmfmac isnt working. Ill start fresh and report back

mattstep commented:
Just as a data point, I tried this on my rpi zero w and the patched firmware resulted in:

`[   33.521314] brcmfmac: brcmf_sdio_bus_rxctl: resumed on timeout
[   33.521349] brcmfmac: brcmf_cfg80211_get_tx_power: error (-52)
[   35.521201] brcmfmac: brcmf_sdio_bus_rxctl: resumed on timeout
[   35.521235] brcmfmac: brcmf_cfg80211_get_tx_power: error (-52)
[   37.521215] brcmfmac: brcmf_proto_bcdc_query_dcmd: brcmf_proto_bcdc_msg failed w/status -110
[   41.521416] brcmfmac: brcmf_dongle_scantime: Scan assoc time error (-52)
[   43.521358] brcmfmac: brcmf_netdev_open: failed to bring up cfg80211
[   45.521306] brcmfmac: brcmf_proto_bcdc_query_dcmd: brcmf_proto_bcdc_msg failed w/status -110
[   45.521346] brcmfmac: brcmf_cfg80211_get_tx_power: error (-52)
[   47.521354] brcmfmac: brcmf_proto_bcdc_query_dcmd: brcmf_proto_bcdc_msg failed w/status -110
[   51.521360] brcmfmac: brcmf_dongle_scantime: Scan assoc time error (-52)
[   53.521352] brcmfmac: brcmf_netdev_open: failed to bring up cfg80211
[   55.531251] brcmfmac: brcmf_proto_bcdc_query_dcmd: brcmf_proto_bcdc_msg failed w/status -110
[   59.531362] brcmfmac: brcmf_dongle_scantime: Scan assoc time error (-52)
[   61.531356] brcmfmac: brcmf_netdev_open: failed to bring up cfg80211
`

NoobieDog commented:
@mattstep , i will be re-doing this for the current git release of NEXMON as i think it will help with some errors we seem to be getting. 

Please bare with me, time is limited right now

mattstep commented:
@NoobieDog No sweat, thanks for your efforts. If I can help at all, let me know.

reedand commented:
Bearing in mind the second-to-last comment but adding this here in case it's useful.  The "Compile a patched firmware" step failed for me with the following output on my Pi 0.

> uname -a
> Linux 4.9.24+ #993 Wed Apr 26 17:56:54 BST 2017 armv6l GNU/Linux 

> ....
> APPLYING FLASHPATCHES gen/flashpatches.mk => brcmfmac43430-sdio.bin (details: log/flashpatches.log)
>   APPLYING PATCHES gen/nexmon.mk => brcmfmac43430-sdio.bin (details: log/patches.log)
>   BUILDING DRIVER  brcmfmac/brcmfmac.ko (details: log/driver.log)
> /home/pi/nexmon/patches/bcm43438/7_45_41_26/nexmon/brcmfmac/cfg80211.c:136:12: error: ‘IEEE80211_BAND_2GHZ’ undeclared here (not in a function)
>   .band   = IEEE80211_BAND_2GHZ,  \
> 
> ...
> 
> /home/pi/nexmon/patches/bcm43438/7_45_41_26/nexmon/brcmfmac/cfg80211.c:6316:7: warning: assignment from incompatible pointer type
>    cap = &wiphy->bands[IEEE80211_BAND_2GHZ]->ht_cap.cap;
>        ^
> make[2]: *** [/home/pi/nexmon/patches/bcm43438/7_45_41_26/nexmon/brcmfmac/cfg80211.o] Error 1
> make[2]: *** Waiting for unfinished jobs....
> make[1]: *** [_module_/home/pi/nexmon/patches/bcm43438/7_45_41_26/nexmon/brcmfmac] Error 2
> Makefile:41: recipe for target 'brcmfmac.ko' failed
> make: *** [brcmfmac.ko] Error 2

Driver.log

> make[1]: Entering directory '/usr/src/linux-headers-4.9.24+'
>   LD      /home/pi/nexmon/patches/bcm43438/7_45_41_26/nexmon/brcmfmac/built-in.o
>   CC [M]  /home/pi/nexmon/patches/bcm43438/7_45_41_26/nexmon/brcmfmac/cfg80211.o
>   CC [M]  /home/pi/nexmon/patches/bcm43438/7_45_41_26/nexmon/brcmfmac/chip.o
> scripts/Makefile.build:293: recipe for target '/home/pi/nexmon/patches/bcm43438/7_45_41_26/nexmon/brcmfmac/cfg80211.o' failed
> Makefile:1490: recipe for target '_module_/home/pi/nexmon/patches/bcm43438/7_45_41_26/nexmon/brcmfmac' failed
> make[1]: Leaving directory '/usr/src/linux-headers-4.9.24+'

wortel538 commented:
I dont want to sound annoying, but when will a fix come for the Zero W? Or if this commit works, any way I can download it with these commits?

matthiasseemoo commented:
I did not have time to work on any Raspberry Pi related stuff in last
couple of weeks. If you want to get nexmon running on a kernel newer than
the one we support, just take the brcmfmac driver from the newer kernel,
change it according to the changes we applied to our brcmfmac driver and
recompile it. Unfortunately, the API changes sometimes between kernel
releases, so you also need to adjust your drivers.

As soon as you are done, just create a pull request to share your changes
with the community.

On Mon, May 22, 2017 at 4:36 PM, wortel538 <notifications@github.com> wrote:

> I dont want to sound annoying, but when will a fix come for the Zero W? Or
> if this commit works, any way I can download it with these commits?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/seemoo-lab/nexmon/pull/55#issuecomment-303118834>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALP_7icGJ7-4qytA6hP0-HFsx64sPM1Lks5r8Z1ugaJpZM4MW2yp>
> .
>



-- 
Matthias Schulz
Secure Mobile Networking Lab - SEEMOO

Email: matthias.schulz@seemoo.tu-darmstadt.de
Web: http://www.seemoo.de/mschulz
Phone (new): +49 6151 16-25478    Fax: +49 6151 16-25471


Department of Computer Science
Center for Advanced Security Research Darmstadt
Technische Universität Darmstadt
Mornewegstr. 32 (Office 4.2.10, Building S4/14)
D-64293 Darmstadt, Germany

NoobieDog commented:
ive finally got some time away from daily work to work on this too, to try more things with this to get it working again.

matthiasseemoo commented:
I would also be interested in having the bcmdhd driver running on the
raspberry pi. the version for the nexus 6p (angler) at least seems to
support monitor mode when compiled without p2p support. but i did not try
it out yet. additionally, the driver has an interface compatible to linux
4.4.

Am 23.05.2017 9:29 nachm. schrieb "NoobieDog" <notifications@github.com>:

> ive finally got some time away from daily work to work on this too, to try
> more things with this to get it working again.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/seemoo-lab/nexmon/pull/55#issuecomment-303506788>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALP_7v_05jTXY3zsiFz2pOz-TvgAQZSDks5r8zOQgaJpZM4MW2yp>
> .
>

DanielAW commented:
see e9d34466e746497ba2ee6951fe2ac58619ac0727, but thanks @NoobieDog

4gordi commented:
Hi NoobieDog, how I can compile and install your fix for pi0w for kali?!?

4gordi commented:
@NoobieDog
I can't install in kali raspberrypi-kernel-headers, without this I can not install nexmon driver(
